sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Shard.cs

```diff
@@ -0,0 +1,184 @@
+            if (async)
+            {
+                await foreach (BlobHierarchyItem blobHierarchyItem in _containerClient.GetBlobsByHierarchyAsync(
+                    prefix: _shardPath).ConfigureAwait(false))
+                {
+                    if (blobHierarchyItem.IsPrefix)
+                        continue;
+
+                    //Chunk chunk = new Chunk(_containerClient, blobHierarchyItem.Blob.Name);
+                    _chunks.Enqueue(blobHierarchyItem.Blob.Name);
+                }
+            }
+            else
+            {
+                foreach (BlobHierarchyItem blobHierarchyItem in _containerClient.GetBlobsByHierarchy(
+                    prefix: _shardPath))
+                {
+                    if (blobHierarchyItem.IsPrefix)
+                        continue;
+
+                    //Chunk chunk = new Chunk(_containerClient, blobHierarchyItem.Blob.Name);
+                    _chunks.Enqueue(blobHierarchyItem.Blob.Name);
+                }
+            }
+
+            // Fast forward to current Chunk
+            if (_chunkIndex > 0)
+            {
+                //TODO possible off by 1 error here.
+                for (int i = 0; i < _chunkIndex; i++)
+                {
+                    _chunks.Dequeue();
+                }
+            }
+
+            _currentChunk = new Chunk(
+                _containerClient,
+                _chunks.Dequeue(),
+                _blockOffset,
+                _eventIndex);
+            _isInitialized = true;
+        }
+
+        public virtual ShardCursor GetCursor()
+            => new ShardCursor(
+                _chunkIndex,
+                _currentChunk.BlockOffset,
+                _currentChunk.EventIndex);
+
+        public virtual bool HasNext()
+        {
+            if (!_isInitialized)
+            {
+                return true;
+            }
+
+            return _chunks.Count > 0 || _currentChunk.HasNext();
+        }
+
+        public virtual async Task<BlobChangeFeedEvent> Next(bool async)
+        {
+            if (!_isInitialized)
+            {
+                await Initalize(async).ConfigureAwait(false);
+            }
+
+            if (!HasNext())
+            {
+                throw new InvalidOperationException("Shard doesn't have any more events");
+            }
+
+            BlobChangeFeedEvent changeFeedEvent;
+
+            changeFeedEvent = await _currentChunk.Next(async).ConfigureAwait(false);
+
+            // Remove currentChunk if it doesn't have another event.
+            if (!_currentChunk.HasNext() && _chunks.Count > 0)
+            {
+                _currentChunk = new Chunk(_containerClient, _chunks.Dequeue());
+                _chunkIndex++;
+            }
+            return changeFeedEvent;
+        }
+
+        /// <inheritdoc/>
+        public void Dispose() => _currentChunk.Dispose();
+
+        /// <summary>
+        /// Constructor for testing.  Do not use.
+        /// </summary>
+        internal Shard(
```

Comment:
These test-only constructors weird me out, but because these object are stateful, it's challenging to get Shard, Chunk, Segment, and Change Feed into the correct starting state for the smaller unit tests.

Reply:
I'd rather mock BlobContainerClient. Over time this might get quite out of sync from "prod" code leading to situation where we're not testing real scenarios.

Reply:
Fixed!